File: services/ai-orchestrator/src/usage.ts
function meteringConfig(): { dbApiUrl: string; internalKey: string } {
    const dbApiUrl = process.env.DB_API_URL?.trim().replace(/\/$/, '');
    const internalKey = process.env.INTERNAL_SERVICE_KEY?.trim();
    if (!dbApiUrl || !internalKey) {
        throw new Error('Usage metering requires DB_API_URL and INTERNAL_SERVICE_KEY');
    }
    return { dbApiUrl, internalKey };
}

export type UsageEvent = {
    tenantId: string; operation: string; success: boolean; tokensIn?: number; tokensOut?: number;
    latencyMs?: number; cacheHit?: boolean; correlationId?: string; projectId?: string;
    snapshotId?: string; runId?: string; costMicrounits?: number; metadata?: Record<string, number | boolean>;
};

export async function checkQuota(tenantId: string): Promise<{ quota_exceeded: boolean; remaining: number; reset_at?: string | null }> {
    if (process.env.METERING_ENABLED === '0') {
        return { quota_exceeded: false, remaining: 999999 };
    }
    
    try {
        const { dbApiUrl, internalKey } = meteringConfig();
        const res = await fetch(`${dbApiUrl}/usage/quota/check?tenant_id=${tenantId}`, {
            headers: {
                'X-Internal-Key': internalKey,
                'X-User-Id': tenantId
            }
        });
        
        if (res.ok) {
            const data = await res.json();
            return {
                quota_exceeded: data.quota_exceeded,
                remaining: data.remaining
            };
        }
    } catch (e) {
        // Fallback: don't break if db is down
    }
    
    return { quota_exceeded: true, remaining: 0 };
}

export async function logUsage(
    tenantId: string, 
    operation: string, 
    success: boolean, 
    tokensIn?: number, 
    tokensOut?: number, 
    latencyMs?: number, 
    cacheHit: boolean = false
): Promise<void> {
    if (process.env.METERING_ENABLED === '0') {
        return;
    }
    
    try {
        const { dbApiUrl, internalKey } = meteringConfig();
        await fetch(`${dbApiUrl}/usage/log`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'X-Internal-Key': internalKey,
                'X-User-Id': tenantId
            },
            body: JSON.stringify({
                tenant_id: tenantId,
                service: 'ai-orchestrator',
                operation,
                success,
                tokens_in: tokensIn,
                tokens_out: tokensOut,
                latency_ms: latencyMs,
                cache_hit: cacheHit
            })
        });
    } catch (e) {
        // Fire and forget, don't crash
    }
}

/** Contract-aligned, bounded telemetry; never accepts prompt/content/secret text. */
export async function logUsageEvent(event: UsageEvent, fetchImpl: typeof fetch = fetch): Promise<void> {
    if (process.env.METERING_ENABLED === '0') return;
    try {
        const { dbApiUrl, internalKey } = meteringConfig();
        await fetchImpl(`${dbApiUrl}/usage/log`, {
            method: 'POST', headers: { 'Content-Type': 'application/json', 'X-Internal-Key': internalKey, 'X-User-Id': event.tenantId },
            body: JSON.stringify({
                tenant_id: event.tenantId, service: 'ai-orchestrator', operation: event.operation, success: event.success,
                tokens_in: event.tokensIn, tokens_out: event.tokensOut, latency_ms: event.latencyMs,
                cache_hit: event.cacheHit ?? false, correlation_id: event.correlationId, project_id: event.projectId,
                snapshot_id: event.snapshotId, run_id: event.runId, cost_microunits: event.costMicrounits,
                event_type: 'usage', status: event.success ? 'completed' : 'failed', metadata: event.metadata ?? {},
            }),
        });
    } catch { /* telemetry is non-critical */ }
}

File: services/document-intelligence/app/usage.py
import os
import httpx
from fastapi import HTTPException
from pydantic import BaseModel
from typing import Optional

DB_API_URL = os.environ.get("DB_API_URL", "http://localhost:8001")

# SECURITY: INTERNAL_SERVICE_KEY must come from environment.
# No hardcoded fallback — fail-closed outside TESTING=1 mode.
_raw_key = os.environ.get("INTERNAL_SERVICE_KEY")
if not _raw_key and os.environ.get("TESTING") != "1":
    raise RuntimeError(
        "INTERNAL_SERVICE_KEY is not set. "
        "Set it via environment before starting document-intelligence. "
        "In test environments, set TESTING=1."
    )
INTERNAL_KEY = _raw_key or "test-internal-key-testing-only"


class QuotaCheckResult(BaseModel):
    quota_exceeded: bool
    remaining: int

async def check_quota(tenant_id: str) -> QuotaCheckResult:
    if os.environ.get("METERING_ENABLED") == "0":
        return QuotaCheckResult(quota_exceeded=False, remaining=999999)
        
    async with httpx.AsyncClient() as client:
        try:
            res = await client.get(
                f"{DB_API_URL}/usage/quota/check",
                params={"tenant_id": tenant_id},
                headers={
                    "X-Internal-Key": INTERNAL_KEY,
                    "X-User-Id": tenant_id
                }
            )
            if res.status_code == 200:
                data = res.json()
                return QuotaCheckResult(
                    quota_exceeded=data.get("quota_exceeded", False),
                    remaining=data.get("remaining", 0)
                )
        except Exception:
            # Fallback allow if DB is unreachable to not break completely
            pass
            
    return QuotaCheckResult(quota_exceeded=False, remaining=999999)

async def log_usage(
    tenant_id: str,
    operation: str,
    success: bool,
    tokens_in: Optional[int] = None,
    tokens_out: Optional[int] = None,
    latency_ms: Optional[int] = None,
    cache_hit: bool = False
):
    if os.environ.get("METERING_ENABLED") == "0":
        return

    async with httpx.AsyncClient() as client:
        try:
            await client.post(
                f"{DB_API_URL}/usage/log",
                json={
                    "tenant_id": tenant_id,
                    "service": "document-intelligence",
                    "operation": operation,
                    "success": success,
                    "tokens_in": tokens_in,
                    "tokens_out": tokens_out,
                    "latency_ms": latency_ms,
                    "cache_hit": cache_hit
                },
                headers={
                    "X-Internal-Key": INTERNAL_KEY,
                    "X-User-Id": tenant_id
                }
            )
        except Exception:
            pass # fire and forget, don't break pipeline on logging failure
